### PR TITLE
PPTP-1525 : WIP : Migrate Registrations

### PR DESCRIPTION
**Pull request by RyanKellyHMRC**
Any existing in-flight registrations which are found to be in the older format are automatically upgraded to the newer format on application startup. Note that this is only performed when the RegistrationUpgrader singleton is created; you will need to run the application in 'production' mode for this to happen.

### CODE REVIEW

File: app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/RegistrationUpgrader.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+          if (incorporationDetails.containsKey(BV_STATUS_OLD))
+            Some(incorporationDetails.getString(BV_STATUS_OLD))
+          else None
+        val registration = incorporationDetails.getDocument(REGISTRATION)
+
+        if (bvStatus.isDefined) {
+          logger.info(s"Upgrading in-flight registration with id [${reg.getString("id")}]")
+          incorporationDetails.remove(BV_STATUS_OLD)
+          registration.append(BV_STATUS_NEW, bvStatus.get)
+          registration.append(IDS_MATCH, BsonBoolean(true)) // Does it matter what we set this to?
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/RegistrationUpgrader.scala:61]: True should be ok here.

File: app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/EagerLoaderModule.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.plasticpackagingtaxregistration.util.migration
+
+import com.google.inject.AbstractModule
+
+class EagerLoaderModule extends AbstractModule {
```
[the remark this answers is not in the document]
Comment (RyanKellyHMRC) [reply app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/EagerLoaderModule.scala:21]: Renamed

File: app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/RegistrationUpgrader.scala (hunk 1 above)
Comment (RyanKellyHMRC) [reply app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/RegistrationUpgrader.scala:61]: Agreed

File: app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/GenericRegistrationRepository.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  private val db         = client.getDatabase(appConfig.mongoDb)
+  private val collection = db.getCollection(appConfig.mongoCollection)
+
+  def upgradeRegistrations(upgrader: Document => Document) =
+    collection.find()
+      .toFuture()
+      .map { registrations =>
+        registrations.foreach { reg =>
+          val updatedReg = upgrader(reg)
+          collection.replaceOne(filter(updatedReg.get[BsonString]("id").get), updatedReg).toFuture()
```
[the remark this answers is not in the document]
Comment (RyanKellyHMRC) [reply app/uk/gov/hmrc/plasticpackagingtaxregistration/util/migration/GenericRegistrationRepository.scala:42]: I had not considered concurrent access due to multiple instances running in production. However, having thought about this, I'm pretty sure that we won't have a problem. Mongo guarantees that all document updates are atomic and all of our migrations are done in the same way, ie. converted from the old format to the new. Even if 1 instance had read a registration just before another had migrated and updated, its update would be identical and therefore not cause an issue.

**Pull request closed without merging**